test/test_jit.py

```diff
@@ -6137,33 +6138,53 @@ def func(a, b):
                             continue
                     msg = ("Failed on {func_name} with inputs {a} {b}. Python: {res_python}, Script: {res_script}"
                            .format(func_name=func_name, a=a, b=b, res_python=res_python, res_script=res_script))
-                    self.assertEqual(res_python, res_script, message=msg, prec=(1e-4) * max(abs(res_python), res_script))
-
-        unary_float_ops = ["log", "log1p", "log10", "exp", "sqrt", "gamma", "lgamma", "erf",
-                           "erfc", "expm1", "fabs", "acos", "asin", "atan", "cos", "sin", "tan",
-                           "asinh", "atanh", "acosh", "sinh", "cosh", "tanh", "degrees", "radians"]
-        binary_float_ops = ["atan2", "fmod", "copysign"]
-        for op in unary_float_ops:
-            checkMathWrap(op, 1)
-        for op in binary_float_ops:
-            checkMathWrap(op, 2)
-
-        checkMath("modf", 1, ret_type="Tuple[float, float]")
-        checkMath("frexp", 1, ret_type="Tuple[float, int]")
-        checkMath("isnan", 1, ret_type="bool")
-        checkMath("isinf", 1, ret_type="bool")
-        checkMath("ldexp", 2, is_float=False, ret_type="float", args_type="(float, int)",
-                  vals=[(i, j) for i in float_vals for j in range(-10, 10)])
-        checkMath("pow", 2, is_float=False, ret_type="int")
-        checkMath("pow", 2, is_float=True, ret_type="float")
-        if not PY2:
-            checkMathWrap("floor", ret_type="int")
-            checkMathWrap("ceil", ret_type="int")
-            checkMathWrap("gcd", 2, is_float=False, ret_type="int")
-            checkMath("isfinite", 1, ret_type="bool")
-        if PY37:
-            checkMathWrap("remainder", 2)
-        checkMathWrap("factorial", 1, is_float=False, ret_type="int", vals=[(i, 0) for i in range(-2, 10)])
+                    if isinstance(res_python, numbers.Number) and isinstance(res_script, numbers.Number):
+                        mx_val = max(abs(res_python), abs(res_script))
+                        prec = 1e-4 * mx_val
+                    else:
+                        prec = (1e-4)
+                    self.assertEqual(res_python, res_script, message=msg, prec=prec)
+
+        unimplemented = ["fsum", "isclose", "trunc", "hypot", "log2"]
+        ops = [x for x in dir(math) if callable(getattr(math, x))]
+        for op in ops:
```

Comment:
Hmmm I somehow feel it's cleaner as old way than this (growing) giant if-else block. cc: @wanchaol @eellison What do you think?

Reply:
well, really, we needed to have a bunch of special cases for calling the different tests anyways. So for each special case for calling the tests, we have an additional if branch, so I don't think it's too bad.

Reply:
I think it's going to be messy no matter what. I guess my general preference is to avoid relying on python implementation details when possible, which we're doing here by calling `dir`. That has been safe so far but who knows what future python versions will do.  (Although we do that in other places too @driazati)

In this case i guess slight preference towards not relying on `dir`, but it's test code and kind of arbitrary and messy so i think it's up to the author here.

Reply:
I think the upside of not needing to manually compare each function in the math documentation 
 with our implementations to see what's implemented is worth the downside of relying on python implementation details. For example, the original issue missed the functions `log2` and `hypot`, presumably because they added them in largely manually.

I also don't think `dir` is really a "python implementation detail". We use it all over the place, and not just in tests.

Reply:
`dir` is picking up attributes that are not public facing methods of the module such as `__loader__`. That feels like implementation details to me. Where else do we use `dir`? I know we do in some places where there isn't an alternative

Reply:
`dir` doesn't get implementation details, everything it gets is well-defined. Ignoring `__whatever__` will leave us with just the public `math` API, which is what we want

Reply:
```
Because dir() is supplied primarily as a convenience for use at an interactive prompt, it tries to supply an interesting set of names more than it tries to supply a rigorously or consistently defined set of names, and its detailed behavior may change across releases. For example, metaclass attributes are not in the result list when the argument is a class.
```

It's not a stable API